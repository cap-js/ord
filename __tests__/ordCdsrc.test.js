const cds = require("@sap/cds");
const path = require("path");

describe("Tests for default ORD document when .cdsrc.json is present", () => {
    let csn, ord;

    beforeAll(async () => {
        cds.root = path.join(__dirname, "bookshop");
        csn = await cds.load(path.join(cds.root, "srv"));
        jest.spyOn(require("../lib/date"), "getRFC3339Date").mockReturnValue("2024-11-04T14:33:25+01:00");
        ord = require("../lib/ord");
    });

    afterAll(() => {
        jest.clearAllMocks();
        jest.resetAllMocks();
    });

    test("Successfully create ORD Documents with defaults", () => {
        const document = ord(csn);
        expect(document).toMatchSnapshot();
    });

    describe("apiResources", () => {
        // eslint-disable-next-line no-useless-escape
        const PACKAGE_ID_REGEX = /^([a-z0-9]+(?:[.][a-z0-9]+)*):(package):([a-zA-Z0-9._\-]+):(v0|v[1-9][0-9]*)$/;

        let document;

        beforeAll(() => {
            document = ord(csn);
        });

        test("PartOfPackage values are valid ORD IDs ", () => {
            for (const apiResource of document.apiResources) {
                expect(apiResource.partOfPackage).toMatch(PACKAGE_ID_REGEX);
            }
        });

        test("The partOfPackage references an existing package", () => {
            for (const apiResource of document.apiResources) {
                expect(
                    document.packages.find(
                        (pck) => pck.ordId === apiResource.partOfPackage
                    )
                ).toBeDefined();
            }
        });
    });

    describe("eventResources", () => {
        // eslint-disable-next-line no-useless-escape
        const GROUP_ID_REGEX = /^([a-z0-9-]+(?:[.][a-z0-9-]+)*):([a-zA-Z0-9._\-/]+):([a-z0-9-]+(?:[.][a-z0-9-]+)*):(?<service>[a-zA-Z0-9._\-/]+)$/;

        let document;

        beforeAll(() => {
            document = ord(csn);
        });

        test("Assigned to exactly one CDS Service group", () => {
            for (const eventResource of document.eventResources) {
                expect(eventResource.partOfGroups.length).toEqual(1);
            }
        });

        test("The CDS Service Group ID includes the CDS Service identifier", () => {
            for (const eventResource of document.eventResources) {
                const [groupId] = eventResource.partOfGroups;
                expect(groupId).toMatch(GROUP_ID_REGEX);

                const match = GROUP_ID_REGEX.exec(groupId);
                if (match && match.groups?.service) {
                    let service = match.groups?.service;
                    if (service.startsWith("undefined"))
                        service = service.replace("undefined.", "");
                    const definition = csn.definitions[service];
                    expect(definition).toBeDefined();
                    expect(definition.kind).toEqual("service");
                }
            }
        });
    });
});
