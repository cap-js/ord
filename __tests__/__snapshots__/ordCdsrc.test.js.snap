// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for default ORD document when .cdsrc.json is present Successfully create ORD Documents with defaults 1`] = `
{
  "$schema": "https://sap.github.io/open-resource-discovery/spec-v1/interfaces/Document.schema.json",
  "apiResources": [
    {
      "apiProtocol": "odata-v4",
      "description": "AdminService",
      "entityTypeMappings": [
        {
          "entityTypeTargets": [
            {
              "ordId": "sap.odm:entityType:BusinessPartner:v1",
            },
          ],
        },
      ],
      "entryPoints": [
        "/odata/v4/admin",
      ],
      "extensible": {
        "supported": "yes",
      },
      "ordId": "sap.test.cdsrc.sample:apiResource:AdminService:v1",
      "partOfPackage": "sap.test.cdsrc.sample:package:capirebookshopordsample-api:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/.well-known/open-resource-discovery/v1/api-metadata/AdminService.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/.well-known/open-resource-discovery/v1/api-metadata/AdminService.edmx",
        },
      ],
      "shortDescription": "AdminService",
      "title": "AdminService",
      "version": "1.0.0",
      "visibility": "public",
    },
    {
      "apiProtocol": "odata-v4",
      "description": "CatalogService",
      "entityTypeMappings": [
        {
          "entityTypeTargets": [],
        },
      ],
      "entryPoints": [
        "/browse",
      ],
      "extensible": {
        "supported": "no",
      },
      "ordId": "sap.test.cdsrc.sample:apiResource:CatalogService:v1",
      "partOfGroups": [
        "sap.cds:service:sap.test.cdsrc.sample:CatalogService",
      ],
      "partOfPackage": "sap.test.cdsrc.sample:package:capirebookshopordsample-api:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/.well-known/open-resource-discovery/v1/api-metadata/CatalogService.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/.well-known/open-resource-discovery/v1/api-metadata/CatalogService.edmx",
        },
      ],
      "shortDescription": "CatalogService",
      "title": "CatalogService",
      "version": "1.0.0",
      "visibility": "public",
    },
  ],
  "dataProducts": [
    {
      "category": "business-object",
      "description": "The Supplier data product offers access to all customers.",
      "entityTypes": [
        "sap.odm:entityType:BusinessPartner:v1",
        "sap.sm:entityType:BusinessPartner:v1",
      ],
      "lastUpdate": "2024-06-20T14:04:01+01:00",
      "localId": "Supplier",
      "ordId": "sap.sm:dataProduct:Supplier:v1",
      "outputPorts": [
        {
          "ordId": "sap.sm:apiResource:SupplierService:v1",
        },
      ],
      "partOfPackage": "sap.sm:package:smDataProducts:v1",
      "releaseStatus": "active",
      "responsible": "sap:ach:CIC-DP-CO",
      "shortDescription": "Ariba Supplier data product",
      "title": "Supplier",
      "type": "primary",
      "version": "1.1.11",
      "visibility": "public",
    },
  ],
  "description": "this is my custom description",
  "eventResources": [
    {
      "description": "CAP Event resource describing events / messages.",
      "extensible": {
        "supported": "no",
      },
      "ordId": "sap.test.cdsrc.sample:eventResource:AdminService:v1",
      "partOfGroups": [
        "sap.cds:service:sap.test.cdsrc.sample:AdminService",
      ],
      "partOfPackage": "sap.test.cdsrc.sample:package:capirebookshopordsample-event:v1",
      "releaseStatus": "beta",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "asyncapi-v2",
          "url": "/.well-known/open-resource-discovery/v1/api-metadata/AdminService.asyncapi2.json",
        },
      ],
      "shortDescription": "Example ODM Event",
      "title": "ODM capirebookshopordsample Events",
      "version": "1.0.0",
      "visibility": "public",
    },
    {
      "description": "CAP Event resource describing events / messages.",
      "extensible": {
        "supported": "no",
      },
      "ordId": "sap.test.cdsrc.sample:eventResource:CatalogService:v1",
      "partOfGroups": [
        "sap.cds:service:sap.test.cdsrc.sample:CatalogService",
      ],
      "partOfPackage": "sap.test.cdsrc.sample:package:capirebookshopordsample-event:v1",
      "releaseStatus": "beta",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "asyncapi-v2",
          "url": "/.well-known/open-resource-discovery/v1/api-metadata/CatalogService.asyncapi2.json",
        },
      ],
      "shortDescription": "Example ODM Event",
      "title": "ODM capirebookshopordsample Events",
      "version": "1.0.0",
      "visibility": "public",
    },
  ],
  "groups": [
    {
      "groupId": "sap.cds:service:sap.test.cdsrc.sample:AdminService",
      "groupTypeId": "sap.cds:service",
      "title": "Admin Service",
    },
    {
      "groupId": "sap.cds:service:sap.test.cdsrc.sample:CatalogService",
      "groupTypeId": "sap.cds:service",
      "title": "Catalog Service",
    },
  ],
  "openResourceDiscovery": "1.9",
  "packages": [
    {
      "description": "Description for capire bookshop ord sample",
      "ordId": "sap.test.cdsrc.sample:package:capirebookshopordsample-api:v1",
      "partOfProducts": [
        "customer:product:capire.bookshop.ord.sample:",
      ],
      "shortDescription": "Short description for capire bookshop ord sample",
      "title": "capire bookshop ord sample",
      "vendor": "customer:vendor:Customer:",
      "version": "1.0.0",
    },
    {
      "description": "Description for capire bookshop ord sample",
      "ordId": "sap.test.cdsrc.sample:package:capirebookshopordsample-event:v1",
      "partOfProducts": [
        "customer:product:capire.bookshop.ord.sample:",
      ],
      "title": "capire bookshop ord sample",
      "vendor": "customer:vendor:Customer:",
      "version": "1.0.1",
    },
    {
      "description": "Description for capire bookshop ord sample version 2",
      "ordId": "sap.test.cdsrc.sample:package:capirebookshopordsample-api:v2",
      "partOfProducts": [
        "customer:product:capire.bookshop.ord.sample:",
      ],
      "shortDescription": "Short description for capire bookshop ord sample version 2",
      "title": "capire bookshop ord sample",
      "vendor": "customer:vendor:Customer:",
      "version": "2.0.0",
    },
  ],
  "policyLevel": "sap:core:v1",
  "products": [
    {
      "ordId": "customer:product:capire.bookshop.ord.sample:",
      "shortDescription": "Description for capire bookshop ord sample",
      "title": "capire bookshop ord sample",
      "vendor": "customer:vendor:customer:",
    },
  ],
}
`;
